recirq/quantum_chess/quantum_board.py

```diff
@@ -311,10 +317,42 @@ def sample(self, num_samples: int) -> List[int]:
             rtn = rtn + samples
         return rtn[:num_samples]
 
-    def _generate_accumulations(self, repetitions: int = 1000) -> None:
+    def _set_full_empty_squares_from_probability(self) -> None:
+        self.full_squares = 0
+        self.empty_squares = 0
+        for i, p in enumerate(self.probabilities):
+            if p == 1:
+                self.full_squares = set_nth_bit(i, self.full_squares, True)
+            if p == 0:
+                self.empty_squares = set_nth_bit(i, self.empty_squares, True)
+
+    def _generate_accumulations(self, repetitions: int = 1000, use_cache: bool = False) -> None:
         """ Samples the state and generates the accumulated 
         probabilities, empty_squares, and full_squares.
         """
+        move_history_cache_key = (tuple(self.move_history), repetitions)
+        move_history_old_cache_key = (tuple(self.move_history[:-1]),
+                                      repetitions)
+        if use_cache and self.move_history:
+            last_move = self.move_history[-1]
+            if move_history_cache_key in self.move_history_probabilities_cache:
+                self.probabilities = self.move_history_probabilities_cache[
+                    move_history_cache_key]
+                self._set_full_empty_squares_from_probability()
+                return
+            cache_key = cache_key_from_move(last_move, repetitions)
+            if move_history_old_cache_key in self.move_history_probabilities_cache and self._caching_supported(last_move) and cache_key in self.cache:
+                probs = self._apply_cache(
+                    self.move_history_probabilities_cache[
+                        move_history_old_cache_key], last_move,
+                    self.cache[cache_key])
+                self.probabilities = probs
+                self._set_full_empty_squares_from_probability()
+                self.move_history_probabilities_cache[
+                    move_history_cache_key] = probs.copy()
+                del self.cache[cache_key]
+                return
+
         self.probabilities = [0] * 64
         self.full_squares = (1 << 64) - 1
```

Comment:
Use the new full squares algorithm here also instead of having 2 implementations

Reply:
Done.